eclair-core/src/main/scala/fr/acinq/eclair/payment/handlers/MultiPartHandler.scala

```diff
@@ -14,35 +14,48 @@
  * limitations under the License.
  */
 
-package fr.acinq.eclair.payment
+package fr.acinq.eclair.payment.handlers
 
-import akka.actor.{Actor, ActorLogging, ActorRef, PoisonPill, Props, Status}
+import akka.actor.{ActorContext, ActorRef, PoisonPill, Status}
+import akka.event.LoggingAdapter
 import fr.acinq.bitcoin.{ByteVector32, Crypto}
 import fr.acinq.eclair.channel.{CMD_FAIL_HTLC, CMD_FULFILL_HTLC, Channel}
-import fr.acinq.eclair.db.{IncomingPayment, IncomingPaymentStatus}
-import fr.acinq.eclair.payment.PaymentLifecycle.ReceivePayment
+import fr.acinq.eclair.db.{IncomingPayment, IncomingPaymentStatus, IncomingPaymentsDb}
+import fr.acinq.eclair.payment.PaymentRequest.ExtraHop
 import fr.acinq.eclair.payment.Relayer.FinalPayload
-import fr.acinq.eclair.wire._
+import fr.acinq.eclair.payment.{MultiPartPaymentHandler, PaymentReceived, PaymentRequest}
+import fr.acinq.eclair.wire.IncorrectOrUnknownPaymentDetails
 import fr.acinq.eclair.{CltvExpiry, MilliSatoshi, NodeParams, randomBytes32}
 
-import scala.concurrent.ExecutionContext
 import scala.util.{Failure, Success, Try}
 
+
 /**
  * Simple payment handler that generates payment requests and fulfills incoming htlcs.
  *
  * Created by PM on 17/06/2016.
  */
-class LocalPaymentHandler(nodeParams: NodeParams) extends Actor with ActorLogging {
+class MultiPartHandler(nodeParams: NodeParams,
+                       db: IncomingPaymentsDb) extends ReceiveHandler {
+
+  import MultiPartHandler._
+
+  // NB: this is safe because this handler will be called from within an actor
+  private var pendingPayments: Map[ByteVector32, (ByteVector32, ActorRef)] = Map.empty
 
-  import LocalPaymentHandler._
+  /**
+   * Can be overridden for a more fine-grained controle of whether or not to handle this payload.
+   * If the call returns false, then the pattern matching will fail and the payload will be passed to other handlers.
+   */
+  def doHandle(p: FinalPayload): Boolean = true
 
-  implicit val ec: ExecutionContext = context.system.dispatcher
-  val paymentDb = nodeParams.db.payments
+  /**
+   * Can be overridden to do custom processing on successfully received payments.
+   */
+  def onSuccess(paymentReceived: PaymentReceived)(implicit log: LoggingAdapter): Unit = ()
```

Comment:
This is currently unused. Not sure yet this is the most elegant way of customizing that, I need to think about it more...